mlflow/server/handlers.py

```diff
@@ -560,14 +599,61 @@ def _search_runs():
 @catch_mlflow_exception
 @_disable_if_artifacts_only
 def _list_artifacts():
+
     request_message = _get_request_message(ListArtifacts())
-    response_message = ListArtifacts.Response()
+
     if request_message.HasField("path"):
         path = request_message.path
     else:
         path = None
+
     run_id = request_message.run_id or request_message.run_uuid
     run = _get_tracking_store().get_run(run_id)
+
+    if _get_serve_artifacts_mode():
+        return _list_artifacts_proxy(run, path)
+    else:
+        return _list_artifacts_non_proxy(run, path)
+
+
+def _append_path_to_artifact_uri(artifact_uri, path):
+
+    parsed_uri = urllib.parse.urlparse(artifact_uri)
+    if parsed_uri.scheme.startswith("http"):
+        _, _, run_path = parsed_uri.path.split("artifacts", 2)
+    else:
+        run_path = parsed_uri.path
+    if path:
+        return posixpath.join(run_path.lstrip("/"), path)
+    else:
+        return run_path.lstrip("/")
+
+
+@catch_mlflow_exception
+def _list_artifacts_proxy(run, path):
+
+    artifact_uri = run.info.artifact_uri
+
+    qualified_path = _append_path_to_artifact_uri(artifact_uri, path)
+
+    artifact_repo = _get_artifact_repo_mlflow_artifacts()
+    files = []
+    for file_info in artifact_repo.list_artifacts(qualified_path):
+        basename = posixpath.basename(file_info.path)
```

Comment:
ListArtifacts always returns paths relative to the run artifact root. This should be `posixpath.join(path, basename)`.